cheatsheets/Server_Side_Request_Forgery_Prevention_Cheat_Sheet.md

```diff
@@ -0,0 +1,387 @@
+
+**Hints for the step 2 regarding the verification on an IP address:**
+
+As mentioned above, not every SDK provide a built-in feature to verify if an IP (V4 + V6) is private/public. So, the following approach can be used based on a blacklist composed of the private IP ranges (*example is given in python in order to be easy to understand and portable to others technologies*) :
+
+```python
+def is_private_ip(ip_address):
+    is_private = False
+    """
+    Determine if a IP address provided is a private one.
+    Return TRUE if it's the case, FALSE otherwise.
+    """
+    # Build the list of IP prefix for V4 and V6 addresses
+    ip_prefix = []
+    # Add prefix for loopback addresses
+    ip_prefix.append("127.")    
+    ip_prefix.append("0.")    
+    # Add IP V4 prefix for private addresses
+    # See https://en.wikipedia.org/wiki/Private_network
+    ip_prefix.append("10.")
+    ip_prefix.append("172.16.")
+    ip_prefix.append("172.17.")
+    ip_prefix.append("172.18.")
+    ip_prefix.append("172.19.")
+    ip_prefix.append("172.20.")
+    ip_prefix.append("172.21.")
+    ip_prefix.append("172.22.")
+    ip_prefix.append("172.23.")
+    ip_prefix.append("172.24.")
+    ip_prefix.append("172.25.")
+    ip_prefix.append("172.26.")
+    ip_prefix.append("172.27.")
+    ip_prefix.append("172.28.")
+    ip_prefix.append("172.29.")
+    ip_prefix.append("172.30.")
+    ip_prefix.append("172.31.")
+    ip_prefix.append("192.168.")
+    ip_prefix.append("169.254.")
+    # Add IP V6 prefix for private addresses
+    # See https://en.wikipedia.org/wiki/Unique_local_address
+    # See https://en.wikipedia.org/wiki/Private_network
+    # See https://simpledns.com/private-ipv6
+    ip_prefix.append("fc")
+    ip_prefix.append("fd")
+    ip_prefix.append("fe")
+    ip_prefix.append("::1")
+    # Verify the provided IP address
+    if ip_address is not None:
+        # Remove whitespace characters from the beginning/end of the string 
+        # and convert it to lower case
+        ip_to_verify = ip_address.strip().lower()
+        # Perform the check against the list of prefix
+        for prefix in ip_prefix:
+            if ip_to_verify.startswith(prefix):
+                is_private = True
+                break
+    return is_private   
+```
+
+### Network layer
+
+Same like for the case n°1.
+
+# References
+
+Online version of the [SSRF bible](https://docs.google.com/document/d/1v1TkWZtrhzRLy0bYXBcdLUedXGb9njTNIJXa3u9akHM) (PDF version is used in this cheat sheet).
+
+Article about [Bypassing SSRF Protection](https://medium.com/@vickieli/bypassing-ssrf-protection-e111ae70727b).
+
+# Tools and code used for schemas
+
+* [Mermaid Online Editor](https://mermaidjs.github.io/mermaid-live-editor) and [Mermaid documentation](https://mermaidjs.github.io/).
+* [Draw.io Online Editor](https://www.draw.io/).
+
+Mermaid code for SSRF common flow (printscreen are used to capture PNG image inserted into this cheat sheet):
+
+```text
```

Comment:
remove text. I don't believe it's a valid one. Everything between 3 backticks is text

Reply:
Taken in account.